scipy/special/_basic.py

```diff
@@ -2639,25 +2636,28 @@ def comb(N, k, exact=False, repetition=False, legacy=True):
     220
 
     """
+    if legacy is not None:
+        warnings.warn(
+            "Using 'legacy' keyword is deprecated and will be removed by "
+            "Scipy 1.13.0. If you want to keep the legacy behaviour, cast "
+            "your inputs directly, e.g. "
+            "'comb(int(your_N), int(your_k), exact=True)'.",
+            DeprecationWarning,
+            stacklevel=2
+        )
     if repetition:
         return comb(N + k - 1, k, exact, legacy=legacy)
     if exact:
-        if int(N) != N or int(k) != k:
-            if legacy:
-                warnings.warn(
-                    "Non-integer arguments are currently being cast to "
-                    "integers when exact=True. This behaviour is "
-                    "deprecated and the default will change to avoid the cast "
-                    "in SciPy 1.11.0. To opt into the future behavior set "
-                    "legacy=False. If you want to keep the argument-casting "
-                    "but silence this warning, cast your inputs directly, "
-                    "e.g. comb(int(your_N), int(your_k), exact=True).",
-                    DeprecationWarning, stacklevel=2
-                )
-            else:
-                return comb(N, k)
-        # _comb_int casts inputs to integers
-        return _comb_int(N, k)
+        if int(N) == N and int(k) == k:
+            # _comb_int casts inputs to integers, which is safe & intended here
+            return _comb_int(N, k)
+        elif legacy:
+            # here at least one number is not an integer; legacy behavior uses
+            # lossy casts to int
+            return _comb_int(N, k)
+        # otherwise, we disregard `exact=True`; it makes no sense for
+        # non-integral arguments
+        return comb(N, k)
```

Comment:
We could start warning about using `exact=True` with non-integers, but I'm fine to discuss this in a separate issue/PR.